### PR TITLE
[SPARK-37137][PYTHON] Inline type hints for python/pyspark/conf.py

### PR DESCRIPTION
**Pull request by ByronHsu**
### What changes were proposed in this pull request?

Inline type hints for python/pyspark/conf.py

### Why are the changes needed?

Currently, Inline type hints for python/pyspark/conf.pyi doesn't support type checking within function bodies. So we inline type hints to support that.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Exising test.

### CODE REVIEW

File: python/pyspark/conf.py [hunk 1]
```diff
@@ -124,48 +130,57 @@ def __init__(self, loadDefaults=True, _jvm=None, _jconf=None):
                 self._jconf = None
                 self._conf = {}
 
-    def set(self, key, value):
+    def set(self, key: str, value: str) -> "SparkConf":
```
[the remark this answers is not in the document]
Comment (zero323) [reply python/pyspark/conf.py:133]: @ueshin @HyukjinKwon WDYT?

[the remark this answers is not in the document]
Comment (ByronHsu) [reply python/pyspark/conf.py:133]: @ueshin  @HyukjinKwon  Could you help me review this patch? Thanks!

[the remark this answers is not in the document]
Comment (HyukjinKwon) [reply python/pyspark/conf.py:133]: Shall we keep it as `Any` for now?

[the remark this answers is not in the document]
Comment (HyukjinKwon) [reply python/pyspark/conf.py:133]: tough call .. I would keep it as just `str` for now though .. for `None` it should be mapped to `null` on JVM ideally.

**Comment by HyukjinKwon** [comment]
add to whitelist

File: python/pyspark/conf.py [hunk 2]
```diff
@@ -178,49 +193,49 @@ def setAll(self, pairs):
...
             if self._jconf is not None:
                 if not self._jconf.contains(key):
                     return None
                 return self._jconf.get(key)
             else:
-                if key not in self._conf:
+                if key not in cast(Dict[str, str], self._conf):
                     return None
-                return self._conf[key]
+                return cast(Dict[str, str], self._conf)[key]
```
Comment (zero323) [inline python/pyspark/conf.py:206]: Unrelated note ‒ shouldn't we use `get` here?

File: python/pyspark/conf.py [hunk 3]
```diff
@@ -124,48 +130,57 @@ def __init__(self, loadDefaults=True, _jvm=None, _jconf=None):
...
 
-    def set(self, key, value):
+    def set(self, key: str, value: str) -> "SparkConf":
         """Set a configuration property."""
         # Try to set self._jconf first if JVM is created, set self._conf if JVM is not created yet.
         if self._jconf is not None:
             self._jconf.set(key, str(value))
         else:
-            self._conf[key] = str(value)
+            cast(Dict[str, str], self._conf)[key] = str(value)
```
Comment (zero323) [inline python/pyspark/conf.py:139]: I'd probably go with `assert` here

```patch
diff --git a/python/pyspark/conf.py b/python/pyspark/conf.py
index 09c8e63d09..a8538b06e4 100644
--- a/python/pyspark/conf.py
+++ b/python/pyspark/conf.py
@@ -136,7 +136,8 @@ class SparkConf(object):
         if self._jconf is not None:
             self._jconf.set(key, str(value))
         else:
-            cast(Dict[str, str], self._conf)[key] = str(value)
+            assert self._conf is not None
+            self._conf[key] = str(value)
         return self
 
     def setIfMissing(self, key: str, value: str) -> "SparkConf":
```

but I guess it is fine for now.

File: python/pyspark/conf.py (hunk 2 above)
Comment (zero323) [reply python/pyspark/conf.py:206]: Also, same as above ‒ single `assert` might be a better option.

**Review by zero323 — APPROVED** [review]
In general, LGTM.

**Comment by ByronHsu** [comment]
@zero323 I have updated the code. Thank you for your precious advice!

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**